servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletOutputStream.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.servlet;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import javax.servlet.ServletOutputStream;
+import javax.servlet.WriteListener;
+
+/**
+ * Servlets output streams (wrapper classes) that control access to the flow.
+ */
+final class DefaultServletOutputStream extends ServletOutputStream {
+    private final DefaultServletHttpResponse response;
+
+    DefaultServletOutputStream(DefaultServletHttpResponse response) {
+        requireNonNull(response, "response");
+        this.response = response;
+    }
+
+    @Override
+    public boolean isReady() {
+        return response.getResponseWriter().isOpen();
+    }
+
+    @Override
+    public void setWriteListener(WriteListener listener) {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public void close() throws IOException {
+        response.flush();
+        response.getResponseWriter().close();
+    }
+
+    @Override
+    public void flush() throws IOException {
+        response.flush();
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        final byte[] bytes = new byte[1];
+        bytes[0] = (byte) b;
+        write(bytes);
```

Comment:
How about using `ByteBuf`s and storing them in an array?
So that we can write when `close()` or `flush()` is called?
```java
private final List<ByteBuf> byteBufs = new ArrayList<>();

...

@Override
public void flush() throws IOException {
    if (!byteBufs.isEmpty()) {
        response.getResponseWriter().write(
                HttpData.wrap(Unpooled.wrappedBuffer(byteBufs.toArray(new ByteBuf[0]))));
        byteBufs.clear();
    }
}

@Override
public void write(int b) throws IOException {
    byteBufs.add(Unpooled.copyInt(b));
}

@Override
public void write(byte[] b, int off, int len) throws IOException {
    byteBufs.add(Unpooled.copiedBuffer(b, off, len));
}

@Override
public void write(byte[] b) throws IOException {
    byteBufs.add(Unpooled.copiedBuffer(b));
}
```

Reply:
I tried your solution but response always failed with exception:
```
com.linecorp.armeria.common.stream.ClosedStreamException
	at com.linecorp.armeria.common.stream.ClosedStreamException.get(ClosedStreamException.java:37)
	at com.linecorp.armeria.common.stream.StreamWriter.write(StreamWriter.java:72)
	at com.linecorp.armeria.server.servlet.ServletPrintWriter.close(ServletPrintWriter.java:58)
	at com.linecorp.armeria.server.servlet.ServletServiceTest$HomeServlet.doGet(ServletServiceTest.java:335)
```

Reply:
I think use ByteArrayOutputStream is good as ByteBuf. Because both append bytes in memory very fast.

Reply:
I tried it in my local and it seems working well. 🤔 
Could you paste your code or let me know what was the problem?

Reply:
My code DefaultServletOutputStream.java : 
```java
final class DefaultServletOutputStream extends ServletOutputStream {
    private final DefaultServletHttpResponse response;
    private final List<ByteBuf> byteBufs = new ArrayList<>();

    DefaultServletOutputStream(DefaultServletHttpResponse response) {
        requireNonNull(response, "response");
        this.response = response;
    }

    @Override
    public boolean isReady() {
        return response.getResponseWriter().isOpen();
    }

    @Override
    public void setWriteListener(WriteListener listener) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void close() throws IOException {
        if (!byteBufs.isEmpty()) {
            response.getResponseWriter().write(
                    HttpData.wrap(Unpooled.wrappedBuffer(byteBufs.toArray(new ByteBuf[0]))));
            byteBufs.clear();
        }
    }

    @Override
    public void flush() throws IOException {
    }

    @Override
    public void write(int b) throws IOException {
        byteBufs.add(Unpooled.copyInt(b));
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
        requireNonNull(b, "b");
        checkArgument(off >= 0, "off: %s (expected: >= 0)", off);
        checkArgument(len >= 0, "len: %s (expected: >= 0)", len);
        byteBufs.add(Unpooled.copiedBuffer(b, off, len));
    }

    @Override
    public void write(byte[] b) throws IOException {
        requireNonNull(b, "b");
        byteBufs.add(Unpooled.copiedBuffer(b));
    }
}
```
ServletServiceTest was failed at doDelete().

Reply:
I think `doDelete()` does not use `DefaultServletOutputStream` because it just calls `response.sendError(HttpStatus.NOT_FOUND.code(), "Not Found");` right away. Did I miss something?

Reply:
No, do you check doDelete() at inner class: ServletServiceTest.HomeServlet (ServletServiceTest.java line 477)

Reply:
Ah there's another `HomeServlet` in the class. Sorry about that I missed. (Perhaps we should rename it to something else to avoid confusion.)
The reason that it didn't work is that it tries to write the data first without writing headers. So I think we should fix that.

If we use `ByteArrayOutputStream`, we have to copy the byte array twice.
But if we use `ByteBuf` we can just copy the data only once. If we change to use the `ByteBufAllocator` from `ServiceRequestContext` later, we can even reduce to allocate additional byte arrays.
So I think it's better to use `ByteBuf`s to hold the data. What do you think?

Reply:
The implementation will be a little tricky so just leave it as it is and fix it later. 😄